File: kickstart/2020-G/d.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <cstring>
using namespace std;

typedef long long ll;
#define MAXN 5000

double dfs(int i, int j, vector<vector<double>> &memo, vector<ll> &presum)
{
    // printf("%d %d\n",i,j);
    if (memo[i][j]>=0)return memo[i][j];
    if (i>=j)return 0;
    
    double res = 0;
    for (int k = i; k < j; k++)
    {
        res += (dfs(i,k,memo,presum)+dfs(k+1,j,memo,presum)+presum[j+1]-presum[i])/(j-i);
    }
    memo[i][j] = res;
    return res;
}

// double solve(vector<int> &arr, int n)
// {
//  vector<vector<double>> memo(MAXN, vector<double>(MAXN,-1));
//  // for(int i=0;i<n;i++)for(int j=0;j<n;j++)printf("%.8f ",memo[i][j]);
//  vector<ll> presum(MAXN,0);
//  for (int i = 0; i < n; i++) presum[i + 1] = presum[i] + arr[i];
//  // printf("%d\n",n);
//  // for (int i = 0; i < n; i++)printf("%d ", arr[i]);
//  // printf("\n");
//  // for (int i = 0; i <= n; i++)printf("%lld ", presum[i]);
//  return dfs(0, n - 1, memo, presum);
//  // return 0;
// }

double solve(vector<int> &arr, int n)
{
    double res = 0;
    for (int mid = 0; mid < n-1; mid++)
    {
        for (int i = 0; i <= mid; i++)res += arr[i] / (double)(mid - i + 1);
        for (int i = mid+1; i < n; i++)res += arr[i] / (double)(i-mid);
    }
    return res;
}

int main()
{
//  freopen("input.txt","r",stdin);
    int T, N;
    scanf("%d", &T);
    vector<int> a(MAXN,0);
    for (int i = 1; i <= T; i++)
    {
        scanf("%d",&N);
        for (int j = 0; j < N; j++)
            scanf("%d", &a[j]);
        double res = solve(a,N);
        printf("Case #%d: %.8f\n", i, res);
    }
    return 0;
}

File: kickstart/2020-G/d.py
import sys
from collections import Counter
from functools import lru_cache
# sys.stdin = open("input.txt", "r")


# def solve(arr):
#     presum = [0]+arr 
#     for i, x in enumerate(presum):
#         if i > 0:
#             presum[i] += presum[i-1]
#     @lru_cache(None)
#     def dfs(i,j):
#         if i == j:
#             return 0
#         res = 0
#         for k in range(i,j):
#             # res += (sum(arr[i:j+1])+dfs(i,k)+dfs(k+1,j))/(j-i)
#             res += (presum[j+1]-presum[i]+dfs(i, k)+dfs(k+1, j))/(j-i)
#         return res
#     return dfs(0,len(arr)-1)

# def solve(arr):
#     n = len(arr)
#     res = 0
#     for mid in range(n-1):
#         for i in range(mid+1):
#             res += arr[i]/(mid-i+1)
#         for i in range(mid+1,n):
#             res += arr[i]/(i-mid)
#     return res 

def solve(arr):
    n = len(arr)
    res, prob = 0, sum(1/i for i in range(1,n))

    for i in range(n):
        res += arr[i]*prob 
        if i < n-1:
            prob -= 1/(n-1-i)
            prob += 1/(i+1)
    return res


t = int(input())
for i in range(t):
    t = int(input())
    arr = list(map(int, input().split()))
    res = solve(arr)
    print("Case #{}: {}".format(i+1, res))
